### ISSUE TITLE
ray.tune.Trainable does not learn when i outsource tensorflow models to a separate class

### ISSUE THREAD

**Issue by MohammadJamali** [ray-project/ray#7492]
I wrote a framework with tensorflow 1.14 in which i will prepare the environment and run the optimizers depend on a config file. it works perfectly when i run it using direct function call.

```
    for module in tf_modules:
        config_copy = copy.deepcopy(config)
        ....
        framework = load_module(config_copy['launcher_config']['structure']['framework'])(config_copy)
        for i in range(epochs):
            print('\r' + framework._train().__str__())
            framework._iteration = i
            print('\r' + name + ' ' + ''.join(['.'] * ((i % 3) + 1)), end='')
        framework._stop()
        del framework
        print('\r {} [Done]'.format(name))
```

```
class TfTrainable(tune.Trainable):
    ...

    def _setup(self, config):
        dataset_loader = load_module(config['dataset']['loader'])
        self.dataset = dataset_loader().load(config)
        self.variables = self._prepare_model(config, self.dataset)
        self.sess = tf.Session()
        self.sess.run(tf.global_variables_initializer())

    def _initialize_variables(self, config, dataset):
        tf_a_index = tf.placeholder(dtype=tf.int32, shape=[None], name='a_index')

        regularizer = tf.contrib.layers.l2_regularizer(scale=config['regularization'])
        initializer = tf.initializers.random_normal(
            mean=config['tensorflow']['random_mean'],
            stddev=config['tensorflow']['random_stddev'],
            seed=config['random_seed'])

        a = tf.get_variable( ... )

        return {
            'tf_a_index': tf_a_index,
            'a': a,
            'a_embedding': tf.nn.embedding_lookup(a, tf_a_index),
             ...
        }

    def _prepare_model(self, config, dataset: dict):
        variables = self._initialize_variables(config, dataset)

        self.modules = []
        for index, module_path in enumerate(config['modules']):
            module = load_module(module_path)()
            self.modules.append(module)
            module.generate_loss(config, variables, dataset)

        return variables

    ...

    def _train(self):
        for module in self.modules:
            module.optimize(self.iteration, {**self.config, 'sess': self.sess},
                            self.variables, self.dataset)

        validation_error = self.compute_errors(self.dataset['validation_groups'], self.variables, self.dataset)
        self.validation_error = self.validation_error.append(validation_error, ignore_index=True)

        return {
            "epoch": self.iteration,
            **validation_error
        }
```

```
class TfSquared(Loss):
    def optimize(self, epoch, config: dict, variables: dict, dataset: dict):
        for batch in slicer(dataset['train'], config['batch_size']):
            config['sess'].run(self.optimizer, feed_dict={
                variables['tf_a_index']: batch[...],
                ...
            })

    def generate_loss(self, config: dict, variables: dict, dataset):
        dot_product = tf.multiply(variables['a'], variables['b'])
        dot_product = tf.reshape(tf.reduce_sum(dot_product, 1, keep_dims=True), (-1,))
        prediction = variables['c'] + dot_product

        loss_function = tf.sqrt(tf.losses.mean_squared_error(variables['tf_true_value'], prediction))

        var_list = [variables['a'], variables['b'], variables['c']]
        self.optimizer = tf.train.GradientDescentOptimizer(learning_rate=config['learning_rate'])
        self.optimizer = self.optimizer.minimize(loss_function, var_list=var_list)

```

Here is the result:

```
'MAP': 0.05866221681687942
'MAP': 0.06426142363467666
'MAP': 0.07129013705117584
...
```

which means it works

but when i run it using `ray.tune` it does not learn anything!
here is the results:

```
tune.run(
            framework,
            name=str(index),
            local_dir=save_dir,
            **config['launcher_config']['tune'],
            scheduler=AsyncHyperBandScheduler(
                **config['launcher_config']['scheduler'],
                **config['launcher_config']['experiment_metric']),
            search_alg=HyperOptSearch(
                {
                    **parameters,
                    **param,
                    **Hyperopt(config['hyperparameters']).space
                },
                max_concurrent=5,
                **config['launcher_config']['experiment_metric']))
```

```
'MAP': 0.05866221681687942
'MAP': 0.05866221681687942
'MAP': 0.05866221681687942
...
```

But it can be fixed if i copy and paste the code inside the loss modules inside the TfTrainable class, but it will break the goal of the written framework.

Can anyone help me fix this problem ?

Thanks a lot

**Comment by richardliaw** [comment]
Can you please provide a reproducible script? Thanks!

**Comment by MohammadJamali** [comment]
After 5 hours of debugging i found out that this is just a simple mistake with random number generator of each thread that i supposed to be same but they weren't. i set that random numbers manually and the problem gone,  sorry for bothering.
Very Very thank you for your time
